subflow.go

```diff
@@ -0,0 +1,248 @@
+			n, err := io.ReadFull(r, buf)
+			if err != nil {
+				pool.Put(buf)
+				sf.close()
+				return
+			}
+			if n != int(sz) {
+				panic(fmt.Sprintf("expect to read %d bytes, read %d", sz, n))
+			}
+			sf.ack(fn)
+			ch <- &frame{fn: fn, bytes: buf}
+			sf.tracker.OnRecv(sz)
+			select {
+			case <-sf.chClose:
+				return
+			default:
+				// continue
+			}
+		}
+	}()
+	probeTimer := time.NewTimer(randomize(probeInterval))
+	for {
+		select {
+		case frame := <-ch:
+			if frame == nil {
+				return
+			}
+			sf.mpc.recvQueue.add(frame)
+			if !probeTimer.Stop() {
+				<-probeTimer.C
+			}
+			probeTimer.Reset(randomize(probeInterval))
+		case <-probeTimer.C:
+			sf.probe()
+			probeTimer.Reset(randomize(probeInterval))
+		}
+	}
+}
+
+func (sf *subflow) sendLoop() {
+	for {
+		select {
+		case <-sf.chClose:
+			return
+		case frame := <-sf.sendQueue:
+			n, err := sf.conn.Write(frame.buf)
+			if err != nil {
+				sf.close()
+				if frame.isDataFrame() {
+					sf.mpc.retransmit(frame)
+				}
+				continue
+			}
+			if n != len(frame.buf) {
+				panic(fmt.Sprintf("expect to write %d bytes on %s, written %d", len(frame.buf), sf.to, n))
+			}
+			if !frame.isDataFrame() {
+				frame.release()
+				continue
+			}
+			if frame.retransmissions == 0 {
+				sf.tracker.OnSent(frame.sz)
+			} else {
+				sf.tracker.OnRetransmit(frame.sz)
+			}
+			log.Tracef("done writing frame %d with %d bytes via %s", frame.fn, frame.sz, sf.to)
+			sf.muPendingAcks.Lock()
+			sf.pendingAcks[frame.fn] = pendingAck{int(frame.sz), time.Now()}
+			sf.muPendingAcks.Unlock()
+			frameCopy := *frame // to avoid race
+			d := sf.retransTimer()
+			time.AfterFunc(d, func() {
+				if sf.isPendingAck(frameCopy.fn) {
+					// No ack means the subflow fails or has a longer RTT
+					sf.updateRTT(d)
+					sf.mpc.retransmit(&frameCopy)
+				} else {
+					// It is ok to release buffer here as the frame will never
+					// be retransmitted again.
+					frame.release()
+				}
+			})
+		}
+	}
+}
+
+func (sf *subflow) ack(fn uint64) {
+	select {
+	case <-sf.chClose:
+		return
+	case sf.sendQueue <- composeFrame(fn, nil):
+	}
+}
+
+func (sf *subflow) gotACK(fn uint64) {
+	if fn == frameTypePing {
+		log.Tracef("pong to %s", sf.to)
+		sf.ack(frameTypePong)
+		return
+	}
+	if fn == frameTypePong {
+		if probeStart := sf.probeStart.Load(); probeStart != nil {
+			start := probeStart.(time.Time)
+			if !start.IsZero() {
+				sf.updateRTT(time.Since(start))
+				sf.probeStart.Store(time.Time{})
+			}
+		}
+		return
+	}
+	log.Tracef("got ack for frame %d from %s", fn, sf.to)
+	sf.muPendingAcks.Lock()
+	defer sf.muPendingAcks.Unlock()
+	pendingAck, found := sf.pendingAcks[fn]
+	if found && pendingAck.sz < maxFrameSizeToCalculateRTT {
+		// it's okay to calculate RTT this way because ack frame is always sent
+		// back through the same subflow, and a data frame is never sent over
+		// the same subflow more than once.
+		sf.updateRTT(time.Since(pendingAck.sentAt))
+		delete(sf.pendingAcks, fn)
```

Comment:
It looks like we don't delete from `sf.pendingAcks` when `pendingAck.sz >= maxFrameSizeToCalculateRTT`, resulting in a memory leak.

Reply:
Nice catch! Will fix.